docs: clarify /healthz and /readyz

zmb3 opened the pull request:
Backports required:
- v7
- v8
- v9

Updates #10799

zmb3 committed:
docs: clarify /healthz and /readyz

Updates #10799

zmb3 commented on docs/pages/setup/admin/troubleshooting.mdx:
```diff
@@ -15,27 +15,9 @@ run with verbose logging enabled by passing it `-d` flag.
   It is not recommended to run Teleport in production with verbose logging as it generates a substantial amount of data.
 </Admonition>
 
-Sometimes you may want to reset [`teleport`](../reference/cli.mdx#teleport) to a clean
```
Removed this section, as it was a duplicate of whats in metrics.mdx.

IMO, the troubleshooting page should be focused on "things aren't working, what steps can I take to fix them" and metrics is more focused on "how can I monitor to ensure things are operating correctly and detect when things start to go wrong."

codingllama commented on docs/pages/setup/reference/metrics.mdx:
```diff
@@ -13,18 +13,68 @@ by default, but you can enable them using the `--diag-addr` flag to `teleport st
+
+Teleport heartbeats run every 5 seconds. This means that depending on the timing
+of heartbeats, it can take 10-20 seconds after connectivity is restored for
+`/readyz` to start reporting healthy again. </Adminition>
```
```suggestion
`/readyz` to start reporting healthy again. </Admonition>
```

ptgott commented on docs/pages/setup/reference/metrics.mdx:
```diff
@@ -13,18 +13,69 @@ by default, but you can enable them using the `--diag-addr` flag to `teleport st
+
+- HTTP 200 OK: Teleport is operating normally
+- HTTP 503 Service Unavailable: Teleport has encountered a connection error and
+  is running in a degraded state. This happens when a Teleport heartbeat fails.
```
Since (as far as I can tell) this is the first time we discuss the heartbeat in the docs, should we introduce the heartbeat at the beginning of the "`/readyz`" section? E.g., which component sends heartbeat traffic to other components and what port and protocol the heartbeat uses.

We cover some of this in the "Recovery" Admonition, but I think it makes sense to introduce this before we first mention the `status` object.

ptgott commented on docs/pages/setup/reference/metrics.mdx:
```diff
@@ -13,18 +13,69 @@ by default, but you can enable them using the `--diag-addr` flag to `teleport st
+The same state information is also available via the `process_state` metric
+under the `/metrics` endpoint.
+
+<Admonition type="note" title="Recovery">
```
I think we can add an H4 section called "Heartbeat" (or similar) to the top of the "`/readyz`" section, then move this text to that H4 section. Then we can add another H4, "Status object" (or similar), which contains the information in "`/readyz`" that isn't in this Admonition. 

I think it would be easier this way for a reader to get familiar with the heartbeat before reading about the status object.

ptgott reviewed (commented):
Since this is the only docs page that mentions the heartbeat, I've recommend separating the heartbeat discussion into a subsection below the "`/readyz`" section to provide the authoritative description of it. 

Is the heartbeat used for more than diagnostics/telemetry (e.g., load balancing)? If it is, it might make sense to talk about the heartbeat in its own page of the docs (e.g., in the Architecture section). 

What do you think?

zmb3 commented:
> Since this is the only docs page that mentions the heartbeat, I've recommend separating the heartbeat discussion into a subsection below the "`/readyz`" section to provide the authoritative description of it.
> 
> Is the heartbeat used for more than diagnostics/telemetry (e.g., load balancing)? If it is, it might make sense to talk about the heartbeat in its own page of the docs (e.g., in the Architecture section).
> 
> What do you think?

If we want to introduce heartbeats separately, it should probably be a separate page, as they're somewhat orthogonal to metrics.

Do you think there's a minimal edit or two we can go here to get this merged and create a separate issue for documenting heartbeats? I'm not a fan of holding up a PR that may provide value now, even if there is still some confusion we need to address.

ptgott commented:
@zmb3 Suggested edits in this PR: https://github.com/gravitational/teleport/pull/11165

ptgott committed:
Edit metrics.mdx (#11165)

- Rename the page, since it's about diagnostics rather than metrics
  alone

- Change major section headings to H2s so they apper in the table of
  contents

- Move information about heartbeats and recovery to an H3 so it's
  more visible

- More information about status codes to a separate H3 below the
  H3 re: heartbeats, both for visibility and to help ensure the reader
  knows the relevant information about heartbeats first

zmb3 committed:
Minor edits

zmb3 committed:
Merge branch 'master' into zmb3/docs-readyz

zmb3 committed:
Merge branch 'master' into zmb3/docs-readyz
